test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,182 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _SORT_BY_KEY_COMMON_H
+#define _SORT_BY_KEY_COMMON_H
+
+#include "support/test_config.h"
+
+#include _PSTL_TEST_HEADER(execution)
+#include _PSTL_TEST_HEADER(algorithm)
+
+#include "support/utils.h"
+
+#include <type_traits>
+
+struct StableSort{};
+struct UnstableSort{};
+
+template<typename _Policy, typename _KeyIt, typename _ValIt, typename _Size, typename StabilityTag>
+void
+call_sort_by_key(_Policy&& policy, _KeyIt keys_begin, _ValIt vals_begin, _Size n, StabilityTag)
+{
+    auto counting_begin = oneapi::dpl::counting_iterator<int>{0};
+
+    // 1. Initialization of buffers
+    std::transform(policy, counting_begin, counting_begin + n, keys_begin,
+                   [](int i) { return i % 2 + 1; });
+    // fill vals_buf with the analogue of std::iota using counting_iterator
+    std::copy(policy, counting_begin, counting_begin + n, vals_begin);
+
+    // 2. Sorting
+    if constexpr (std::is_same_v<StabilityTag, StableSort>)
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+}
+
+template<typename _Keys, typename _Vals, typename _Size>
+void
+check_sort_by_key_results(const _Keys& keys_buf, const _Vals& vals_buf,  _Size n)
```

Comment:
There do not seem to be any checks for the stability of sorting; I think it is needed.

Reply:
Agree. I've added the check. It's revealed that the host-side implementation is not stable:

https://github.com/oneapi-src/oneDPL/blob/e81af7bc5ee7a40ef28040b9d6aa405bbdbf1223/include/oneapi/dpl/pstl/algorithm_impl.h#L2478-L2499

Tagging @danhoeflinger for awareness.

The documentation says that `sort_by_key` is stable. But host-side policies implement non-stable one. I see 2 ways to handle it:

1. Treat it as an implementation bug. Fix it.
2. Treat it as a documentation bug for host-side policies. Fix the documentation and introduce a stable host-side implementation.

I see the 1st approach as a better way to proceed, because some users may already have `sort_by_key` assuming stability. @akukanov, what is your opinion on that?

Reply:
I've partially followed the 2nd approach. `sort_by_key` does not guarantee the stability now, but there is a exception: device execution policies.